src/bridged_logging.rs

```diff
@@ -0,0 +1,226 @@
+#[derive(Debug, Copy, Clone)]
+pub enum GlibLoggerFormat {
+    /// A simple format, writing only the message on output.
+    Plain,
+    /// A simple format, writing file, line and message on output.
+    LineAndFile,
+    /// A logger using glib structured logging.
+    Structured,
+}
+
+/// Enumeration of the possible domain handling behaviours for a [`GlibLogger`]
+pub enum GlibLoggerDomain {
+    /// Logs will have no domain specified.
+    None,
+    /// Logs will use the `target` of the log crate as a domain; this allows
+    /// Rust code like `warn!(target: "my-domain", "...");` to log to the glib
+    /// logger using the specified domain.
+    CrateTarget,
+    /// Logs will use the crate path as the log domain.
+    CratePath,
+    /// Logs will use the specified domain for logging. Note that this option
+    /// due to the way the `log` crate is implemented, is global for all crates
+    /// in the current process.
+    Custom(&'static str),
+}
+
+/// An implementation of a `log` crate logger which logs over glib logging
+/// facilities.
+///
+/// Example:
+///
+/// ```no_run
+/// static glib_logger: glib::log_bridge::GlibLogger = glib::log_bridge::GlibLogger {
+///     format: glib::log_bridge::GlibLoggerFormat::Structured,
+///     domain: glib::log_bridge::GlibLoggerDomain::CrateTarget,
+/// };
+///
+/// fn main() {
+///     log::set_logger(&glib_logger);
+///     log::set_max_level(log::LevelFilter::Debug);
+///
+///     info!("This line will get logged by glib");
+/// }
+/// ```
+pub struct GlibLogger {
+    pub format: GlibLoggerFormat,
+    pub domain: GlibLoggerDomain,
+}
+
+impl GlibLogger {
+    fn level_to_glib(level: rs_log::Level) -> glib_sys::GLogLevelFlags {
+        match level {
+            // Errors are mapped to critical to avoid automatic termination
+            rs_log::Level::Error => glib_sys::G_LOG_LEVEL_CRITICAL,
+            rs_log::Level::Warn => glib_sys::G_LOG_LEVEL_WARNING,
+            rs_log::Level::Info => glib_sys::G_LOG_LEVEL_INFO,
+            rs_log::Level::Debug => glib_sys::G_LOG_LEVEL_DEBUG,
+            // There is no equivalent to trace level in glib
+            rs_log::Level::Trace => glib_sys::G_LOG_LEVEL_DEBUG,
+        }
+    }
+
+    fn write_log(domain: Option<&str>, level: rs_log::Level, message: &str) {
+        unsafe {
+            crate::glib_sys::g_log(
+                domain.to_glib_none().0,
+                GlibLogger::level_to_glib(level),
+                message.replace("%", "%%").to_glib_none().0,
+            );
+        }
+    }
+
+    fn write_log_structured(
+        domain: Option<&str>,
+        level: log::Level,
+        file: Option<&str>,
+        line: Option<u32>,
+        func: Option<&str>,
+        message: &str
+    ) {
+        let line_str = match line {
+            None => None,
+            Some(l) => Some(l.to_string()),
+        };
+
+        unsafe {
+            crate::glib_sys::g_log_structured_standard(
```

Comment:
The structured logging feature is new in GLib 2.50, but as we don't actually log any custom structured information just going via `g_log` should be sufficient or not?

Reply:
Oh nevermind, actually we do. And this function is new in 2.56 :) So needs some condition compilation around the `v2_56` feature.